### PR TITLE
Add tflint as built-in hook

### PR DESCRIPTION
**Pull request by marinalimeira**
## Description

This task is part of https://github.com/gruntwork-io/terraform-aws-cis-service-catalog/issues/486, a repo only accessible to Gruntwork customers.

The team is taking the direction of adding `tflint` to Terragrunt as part of the CIS v1.5 upgrade. It will enable us to validate the `inputs` to give a feedback to the user before any resource is deployed. The focus now is to validate the security group rules.

⚠️ [The implementation of `tflint` is attached to the CLI](https://github.com/terraform-linters/tflint/blob/eefb06ad833b72eaca8c68b57a0557cac431d187/cmd/inspect.go#LL16), so I couldn't find an quick way to use their interface as a library. 

⚠️ **The binary size increased from 33MB to 39MB.**

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [ ] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.
- [ ] How to not break projects that might already have tflint as a hook?
- [x] `inputs` should be parsed as variables to tflint, we need to validate these variables also.
- [x] [Update license for the files that use tflint.](https://www.mozilla.org/en-US/MPL/headers/)
- [x] Use config file (.tflint.hcl) from one of the parent folders.
- [x] Add docs about how the tflint works.
- [x] Integration test works.

## Release Notes (draft)

Added tflint as built-in hook.

### CODE REVIEW

**Comment by robmorgan** [comment]
@marinalimeira great! Some questions from me:

1. Does it leave any `tflint` processes hanging around after `terragrunt` finishes executing?
2. Do we need to hit up our lawyers to check the license compatibility and how we would disclose that in this repo?

**Comment by denis256** [comment]
Conflicting files go.mod / go.sum

Looks like we need to upgrade Go version to 1.18:
```
Stderr: # golang.org/x/exp/constraints
../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:13:2: syntax error: unexpected ~, expecting method or interface name
../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:20:2: syntax error: unexpected ~, expecting method or interface name
../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:27:9: syntax error: unexpected |, expecting semicolon or newline or }
../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:34:2: syntax error: unexpected ~, expecting method or interface name
../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:41:2: syntax error: unexpected ~, expecting method or interface name
../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:49:10: syntax error: unexpected |, expecting semicolon or newline or }
note: module requires Go 1.18
```

**Comment by marinalimeira** [comment]
> Conflicting files go.mod / go.sum
> 
> Looks like we need to upgrade Go version to 1.18:
> 
> ```
> Stderr: # golang.org/x/exp/constraints
> ../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:13:2: syntax error: unexpected ~, expecting method or interface name
> ../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:20:2: syntax error: unexpected ~, expecting method or interface name
> ../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:27:9: syntax error: unexpected |, expecting semicolon or newline or }
> ../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:34:2: syntax error: unexpected ~, expecting method or interface name
> ../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:41:2: syntax error: unexpected ~, expecting method or interface name
> ../go/pkg/mod/golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e/constraints/constraints.go:49:10: syntax error: unexpected |, expecting semicolon or newline or }
> note: module requires Go 1.18
> ```

I saw that you already have a draft PR @denis256 ! This PR is not critical at the moment, we are still validating other things, so I can wait for that one, thanks.

**Comment by marinalimeira** [comment]
Analysis of the error message `Failed to initialize plugins; Unrecognized remote plugin message: Incorrect Usage. flag provided but not defined: -act-as-bundled-plugin`:

When the `terraform` plugin is definied [without `version` or without `source`](https://github.com/terraform-linters/tflint/blob/78315a6fbf3626140affb3fb237164dae40204d6/plugin/install.go#L37), it executes `self` with `--act-as-bundled-plugin` ([source](https://github.com/terraform-linters/tflint/blob/eefb06ad833b72eaca8c68b57a0557cac431d187/plugin/discovery.go#LL38C56-L38C56)):
```go
cmd = exec.Command(self, "--act-as-bundled-plugin")
```

When `tflint` runs inside Terragrunt, it ends up running `terragrunt --act-as-bundled-plugin`. `Failed to initialize plugins; Unrecognized remote plugin message:` is an error message from `tflint` and ` Incorrect Usage. flag provided but not defined: -act-as-bundled-plugin` is an error message from Terragrunt.

In order to fix this, the `.tflint.hcl` file needs `source` and/or `version:
```
plugin "terraform" {
  enabled = true
  preset  = "recommended"

  version = "0.1.0"
  source  = "github.com/terraform-linters/tflint-ruleset-terraform"
}
```

Without this `.tflint.hcl` file, the `tflint` command inside Terragrunt won't work, given it can't call inside terragrunt the function `exec.Command(self, "--act-as-bundled-plugin")`.

**Comment by denis256** [comment]
Merge conflicts need to be fixed

**Comment by marinalimeira** [comment]
Current status: 
- merge conflicts fixed
- started integration tests, failing so I am investigating why (manual test with terragrunt plan works)

@denis256 this is ready for a first review! Thanks

File: cli/aws_provider_patch.go [hunk 1]
```diff
@@ -285,24 +284,28 @@ func overrideAttributeInBlock(block *hclwrite.Block, key string, value string) (
 //
 // Assume block is:
 //
-// provider "aws" {
-//   region = var.aws_region
-//   assume_role {
-//     role_arn = var.role_arn
-//   }
-// }
+//	provider "aws" {
```
Comment (brikis98) [inline cli/aws_provider_patch.go:287]: Can you undo all these whitespace changes?

File: cli/cli_app.go [hunk 2]
```diff
@@ -715,16 +717,27 @@ func processHooks(hooks []config.Hook, terragruntOptions *options.TerragruntOpti
...
-				terragruntOptions.Logger.Errorf("Error running hook %s with message: %s", curHook.Name, possibleError.Error())
-				errorsOccured = multierror.Append(errorsOccured, possibleError)
+
+			if actionToExecute == "tflint" {
+				actionParams := curHook.Execute[0:]
+
+				err := tflint.RunTflintWithOpts(terragruntOptions, terragruntConfig, actionParams)
+				if err != nil {
+					terragruntOptions.Logger.Errorf("Error running hook %s with message: %s", curHook.Name, err.Error())
+					multierror.Append(errorsOccured, err)
```
Comment (brikis98) [inline cli/cli_app.go:727]: This should be `errorsOccurred = multierror.Append(...)`. See the analogous code in the else statement.

File: cli/cli_app.go [hunk 3]
```diff
@@ -1057,7 +1070,7 @@ func providersNeedInit(terragruntOptions *options.TerragruntOptions) bool {
 //
 // If terraformSource is specified, then arguments to download the terraform source will be appended to the init command.
 //
-// This method will return an error and NOT run terraform init if the user has disabled Auto-Init
+// # This method will return an error and NOT run terraform init if the user has disabled Auto-Init
```
Comment (brikis98) [inline cli/cli_app.go:1073]: Please revert all unnecessary changes to comments in this PR.

File: cli/tfsource/types.go [hunk 4]
```diff
@@ -111,25 +111,25 @@ func (terraformSource TerraformSource) WriteVersionFile() error {
...
+//     method). We also assume the version of the module is stored in the query string (e.g. ref=v0.0.3), so we store
+//     the base 64 encoded sha1 of the query string in a file called .terragrunt-source-version within /T/W/H.
 //
 // The downloadTerraformSourceIfNecessary decides when we should download the Terraform code and when not to. It uses
 // the following rules:
 //
-// 1. Always download source URLs pointing to local file paths.
-// 2. Only download source URLs pointing to remote paths if /T/W/H doesn't already exist or, if it does exist, if the
-//    version number in /T/W/H/.terragrunt-source-version doesn't match the current version.
+//  1. Always download source URLs pointing to local file paths.
```
Comment (brikis98) [inline cli/tfsource/types.go:130]: And more comment changes to revert...

File: test/fixture-validate-inputs/fail-unused-varfile/varfiles/main.tfvars [hunk 5]
```diff
@@ -1,2 +1,2 @@
-input = "hello world"
+input        = "hello world"
```
Comment (brikis98) [inline test/fixture-validate-inputs/fail-unused-varfile/varfiles/main.tfvars:1]: Is this an intentional change?

File: tflint/README.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+talk about license
+what is this
+where is being used
+# tflint
```
Comment (brikis98) [inline tflint/README.md:4]: TODO

File: tflint/tflint.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	if statusCode != 0 {
+		errorMsg := fmt.Sprintf("Error while running 'tflint'! Status code: %d", statusCode)
+		return errors.New(errorMsg)
+	}
+
+	args = append(args, variables...)
+	args = append(args, terragruntOptions.WorkingDir)
+
+	terragruntOptions.Logger.Debugf("Running tflint with args %v", args)
+	statusCode = cli.Run(args)
```
Comment (brikis98) [inline tflint/tflint.go:32]: Does `tflint` logging show up OK in `stdout` / `stderr` when you run this?

File: tflint/tflint.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+func inputsToTflintVar(inputs map[string]interface{}) ([]string, error) {
+	var variables []string
+	for key, value := range inputs {
+		varValue, err := util.AsTerraformEnvVarJsonValue(value)
+		if err != nil {
+			return nil, err
+		}
+
+		newVar := fmt.Sprintf("--var=%s=%s", key, varValue)
```
Comment (brikis98) [inline tflint/tflint.go:50]: We need a test that checks what happens when this when the variable value is a list, a string, a string with whitespace, a number, and so on. As in, do they get passed properly as CLI params and does `tflint` parse them correctly?

**Review by brikis98 — COMMENTED** [review]
1. Please update the docs.
2. Please add more tests for different kinds of tflint checks, including those expected to fail.

File: cli/cli_app.go (hunk 2 above)
Comment (marinalimeira) [reply cli/cli_app.go:727]: Good catch! Thanks

File: cli/cli_app.go (hunk 3 above)
Comment (marinalimeira) [reply cli/cli_app.go]: Fixed [f334d25](https://github.com/gruntwork-io/terragrunt/pull/2362/commits/f334d25d508af32dd3c0fac2fe68450cb3fd2628) . My pre-commit locally is broken, so the quick fix was to run `goimports -l -w .`... it changed unrelated things.

File: test/fixture-validate-inputs/fail-unused-varfile/varfiles/main.tfvars (hunk 5 above)
Comment (marinalimeira) [reply test/fixture-validate-inputs/fail-unused-varfile/varfiles/main.tfvars:1]: No it wasnt

File: tflint/tflint.go (hunk 8 above)
Comment (marinalimeira) [reply tflint/tflint.go:64]: Started some tests here: https://github.com/gruntwork-io/terragrunt/blob/ff46a2460d08aed5c3b4a8b76ec7357a40905f11/util/jsons_test.go

Need to add these other scenarios

**Comment by marinalimeira** [comment]
I am working now to understand why `TestTflintFindsIssuesWithInvalidInput` is failing, but this might be carried over for tomorrow. It works locally and I can't understand why it fails in the CI.

In the meanwhile, @brikis98 @denis256 this feature is not WIP anymore. Can you please take a look?

File: tflint/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+```hcl
+terraform {
+  before_hook "before_hook" {
+    commands     = ["apply", "plan"]
+    execute      = ["tflint"]
+  }
+}
+```
+
+The `.tflint.hcl` should exist in the same folder or one of it's parents. If Terragrunt can't find a `.tflint.hcl` file, it won't execute tflint.
```
Comment (ina-stoyanova) [inline tflint/README.md:19]: Which "same" folder? I imagine it's at the root level.

File: tflint/README.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+plugin "aws" {
+    enabled = true
+    version = "0.21.0"
+    source  = "github.com/terraform-linters/tflint-ruleset-aws"
+}
+```
+
+### Configuration
+
+The `execute` parameter only accepts `tflint`, it will ignore any other parameter. Any desired extra configuration should be added in the `.tflint.hcl` file. It will work with a `.tflint.hcl` file in any parent folder.
```
Comment (ina-stoyanova) [inline tflint/README.md:30]: NIT: The docs look really neat! 🙌 It might also be really helpful to have an example folder structure shown here to demonstrate what you mean by "parent" folder, where a module would live, and where the tflint configs would live

File: tflint/README.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
+# tflint 
+
+This package is a wrapper of the ![tflint](https://github.com/terraform-linters/tflint) project, enabling it to be natively executed from the before and after hooks. The `tflint/tflint.go` uses the MPL 2.0 license, due to the license of tflint.
```
Comment (brikis98) [inline tflint/README.md:3]:
```suggestion
This package allows us to embed [tflint](https://github.com/terraform-linters/tflint) in Terragrunt, enabling it to be natively executed from the before and after hooks without having to install `tflint` separately. Since `tflint` is licensed with MPL, we are required to let you know where you can find its source code: https://github.com/terraform-linters/tflint
```

Per Q8 of MPL, as far as I can tell, using `tflint` this way (unmodified) does not change the license of Terragrunt. It just requires us to tell users where they can find the source code of the MPL project: https://www.mozilla.org/en-US/MPL/2.0/FAQ/

File: tflint/README.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+## How to use
```
Comment (brikis98) [inline tflint/README.md:5]: These docs don't belong here. They belong in the docs site. See the `docs` folder.

File: tflint/README.md (hunk 9 above)
Comment (brikis98) [reply tflint/README.md:19]: If no `.tflint.hcl` file is found, does it show an error? Or just exit? Is there any logging to explain what happened?

File: tflint/tflint.go [hunk 13]
```diff
@@ -0,0 +1,117 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
```
Comment (brikis98) [inline tflint/tflint.go:3]: I think you should instead say here that this code embeds `tflint`, which is under an MPL license, and you can find its source code at `<URL>`.

File: tflint/tflint.go [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	args = append(args, terragruntOptions.WorkingDir)
+
+	terragruntOptions.Logger.Debugf("Running tflint with args %v", args)
+	statusCode = cli.Run(args)
+
+	if statusCode == cmd.ExitCodeError {
+		return errors.WithStackTrace(ErrorRunningTflint{args: initArgs})
+	} else if statusCode == cmd.ExitCodeIssuesFound {
+		// When tflint finds issues in the project, it prints out but does not stop execution of Terragrunt.
+		terragruntOptions.Logger.Warnf("Tflint found issues in the project. Check for the tflint logs.")
```
Comment (brikis98) [inline tflint/tflint.go:52]: If tflint returned an error, shouldn't we return an error?

File: tflint/tflint.go [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+	} else if statusCode == cmd.ExitCodeOK {
+		terragruntOptions.Logger.Info("Tflint has run successfully. No issues found.")
+	} else {
+		terragruntOptions.Logger.Warnf("Unknown status code from tflint: %d", statusCode)
```
Comment (brikis98) [inline tflint/tflint.go:56]: Shouldn't this return an error?

File: tflint/tflint.go [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+	return variables, nil
+}
+
+// findTflintConfigInProjects looks for a .tflint.hcl file in the current folder or it's parents.
+func findTflintConfigInProject(terragruntOptions *options.TerragruntOptions) (string, error) {
+	previousDir := terragruntOptions.WorkingDir
+
+	// To avoid getting into an accidental infinite loop (e.g. do to cyclical symlinks), set a max on the number of
+	// parent folders we'll check
+	for i := 0; i < terragruntOptions.MaxFoldersToCheck; i++ {
```
Comment (brikis98) [inline tflint/tflint.go:82]: NIT: seems like a common "find in parent folders" method could be extracted from this code, rather than copy pasted.

**Review by robmorgan — COMMENTED** [review]
@marinalimeira LGTM. I think it's worth addressing most of Jim's suggestions if you have time. One other thing - Is it mentioned anywhere that the user needs to install `tflint` if they want to use the hook? Maybe we need a few sentences in the `README.md` or docs site and a minimum suggested version they should install.

**Comment by robmorgan** [comment]
@marinalimeira additionally a 'How to Test' section would be super valuable. If I try something manually, e.g:

```bash
cd test/fixture-tflint/issues-found
terragrunt apply -auto-approve
```

It executes the hook but applies successfully. I'm expecting anything in a directory called `issues-found` to fail. Should it be returning an error?

File: tflint/tflint.go (hunk 13 above)
Comment (marinalimeira) [reply tflint/tflint.go]: Updated 6e09bc8

File: test/integration_test.go [hunk 17]
```diff
@@ -5184,3 +5200,36 @@ func validateBoolOutput(t *testing.T, outputs map[string]TerraformOutput, key st
...
+		t.Fatal("tflint execution failed")
+	}
+}
+
+func TestTflintFindsIssuesWithInvalidInput(t *testing.T) {
+	errOut := new(bytes.Buffer)
+	rootPath := copyEnvironmentWithTflint(t, TEST_FIXTURE_TFLINT_ISSUES_FOUND)
+	modulePath := util.JoinPath(rootPath, TEST_FIXTURE_TFLINT_ISSUES_FOUND)
+	err := runTerragruntCommand(t, fmt.Sprintf("terragrunt plan --terragrunt-working-dir %s", modulePath), os.Stdout, errOut)
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (marinalimeira) [reply test/integration_test.go:5221]: Fixed! 6394523

File: tflint/tflint.go (hunk 14 above)
Comment (marinalimeira) [reply tflint/tflint.go:51]: Fixed 6394523

File: tflint/tflint.go (hunk 15 above)
Comment (marinalimeira) [reply tflint/tflint.go:55]: Fixed d7a1835

**Comment by marinalimeira** [comment]
> @marinalimeira additionally a 'How to Test' section would be super valuable. If I try something manually, e.g:
> 
> ```shell
> cd test/fixture-tflint/issues-found
> terragrunt apply -auto-approve
> ```
> 
> It executes the hook but applies successfully. I'm expecting anything in a directory called `issues-found` to fail. Should it be returning an error?

Fixed this! 6394523 I am returning an error now.

**Review by robmorgan — APPROVED** [review]
